rust-runtime/aws-smithy-http-server/src/operation/mod.rs

```diff
@@ -0,0 +1,268 @@
+//!     todo!()
+//! }
+//! ```
+//!
+//! ## [`OperationService`]
+//!
+//! Similarly, the [`OperationService`] trait is implemented by all `Service<(Op::Input, ...)>` with
+//! `Response = Op::Output`, and `Error = OperationError<Op::Output, PollError>`.
+//!
+//! We use [`OperationError`], with a `PollError` not constrained by the model, to allow the user to provide a custom
+//! [`Service::poll_ready`](tower::Service::poll_ready) implementation.
+//!
+//! The following are examples of [`Service`](tower::Service)s which implement [`OperationService`]:
+//!
+//! - `Service<CartIdentifier, Response = ShoppingCart, Error = Infallible>`.
+//! - `Service<(CartIdentifier, Extension<Context>), Response = ShoppingCart, Error = GetShoppingCartError>`.
+//! - `Service<(CartIdentifier, Extension<Context>, Extension<ExtraContext>), Response = ShoppingCart, Error = GetShoppingCartError)`.
+//!
+//! Notice the parallels between [`OperationService`] and [`Handler`].
+//!
+//! ## Constructing an [`Operation`]
+//!
+//! The following is an example of using both construction approaches:
+//!
+//! ```rust
+//! # use std::task::{Poll, Context};
+//! # use aws_smithy_http_server::operation::*;
+//! # use tower::Service;
+//! # pub struct CartIdentifier;
+//! # pub struct ShoppingCart;
+//! # pub enum GetShoppingError {}
+//! # pub struct GetShopping;
+//! # impl OperationShape for GetShopping {
+//! #    const NAME: &'static str = "GetShopping";
+//! #
+//! #    type Input = CartIdentifier;
+//! #    type Output = ShoppingCart;
+//! #    type Error = GetShoppingError;
+//! # }
+//! # type OpFuture = std::future::Ready<Result<ShoppingCart, OperationError<GetShoppingError, PollError>>>;
+//! // Construction of an `Operation` from a `Handler`.
+//!
+//! async fn op_handler(input: CartIdentifier) -> Result<ShoppingCart, GetShoppingError> {
+//!     todo!()
+//! }
+//!
+//! let operation = GetShopping::from_handler(op_handler);
+//!
+//! // Construction of an `Operation` from a `Service`.
+//!
+//! pub struct PollError;
+//!
+//! pub struct OpService;
+//!
+//! impl Service<(CartIdentifier, ())> for OpService {
+//!     type Response = ShoppingCart;
+//!     type Error = OperationError<GetShoppingError, PollError>;
+//!     type Future = OpFuture;
+//!
+//!     fn poll_ready(&mut self, cx: &mut Context) -> Poll<Result<(), Self::Error>> {
+//!         // NOTE: This MUST NOT return `Err(OperationError::Model(_))`.
+//!         todo!()
+//!     }
+//!
+//!     fn call(&mut self, request: (CartIdentifier, ())) -> Self::Future {
+//!         // NOTE: This MUST NOT return `Err(OperationError::Poll(_))`.
+//!         todo!()
+//!     }
+//! }
+//!
+//! let operation = GetShopping::from_service(OpService);
+//!
+//! ```
+//!
+//! ## Upgrading Smithy services to HTTP services
+//!
+//! Both [`Handler`] and [`OperationService`] accept and return Smithy model structures. After an [`Operation`] is
+//! constructed they are converted to a normal form
```

Comment:
```suggestion
//! constructed they are converted to a canonical form
```

I'd say in CS most people associate "normal form" with [these](https://en.wikipedia.org/wiki/Database_normalization#Normal_forms).